File: Geeks4Geeks/longest_common_subsequence.py
"""
 * Author Sushant Moon
 * Email sushant.moon@gmail.com
 * Create date 2018-08-15 08:08:37
 * Modify date 2018-08-15 08:08:37
 * Description Longest common subsequence (https://practice.geeksforgeeks.org/problems/longest-common-subsequence/0/)
"""

def lcs(a,b,la,lb):
    l = [[None]*la for i in range(lb)]
    for i in range(lb):
        for j in range(la):
            if b[i] == a[j]:
                if ((i == 0) or (j == 0)):
                    l[i][j] = 1
                else:
                    l[i][j] = 1 + l[i-1][j-1]
            else:
                top = 0
                left = 0
                if (i > 0):
                    top = l[i-1][j]
                if (j > 0):
                    left = l[i][j-1]
                l[i][j] = max(top, left)
    # print(l)
    return l[lb-1][la-1]

def main():
    t = int(input().strip())
    while(t>0):
        j = input()
        [la, lb] = [int(x.strip()) for x in j.split()]
        a = input()
        b = input()
        print(lcs(a,b,la,lb))
        t -= 1
main()

File: Geeks4Geeks/longest_common_subsequence.c
/*
* =====================================================================================
*
*       Filename:  longest_common_subsequence.c
*
*    Description:  https://practice.geeksforgeeks.org/problems/longest-common-subsequence/0
*
*        Version:  1.0
*        Created:  03/15/2019 03:31:44 AM
*       Revision:  none
*       Compiler:  gcc
*
*         Author:  Sushant Moon (), sushant.moon@gmail.com
*   Organization:  
*
* =====================================================================================
*/


#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define max(a,b) (a>b)?a:b


int longest_common_subsequence(char *A, char *B, int len_A, int len_B){
	int i, j;

	// defining M
	int **M = (int **)malloc(sizeof(int *)*(len_A+1));
	for(i = 0; i<=len_A; i++){
		M[i] = (int *)malloc(sizeof(int)*(len_B+1));
	}

	// filling up usgin recurrence
	for(i=0; i<=len_A; i++){
		for(j=0; j<=len_B; j++){
			if(i == 0 || j == 0){
				M[i][j] = 0;
			}
			else if(A[i-1] == B[j-1]){
				M[i][j] = M[i-1][j-1] + 1;
			}
			else{
				M[i][j] = max(M[i-1][j], M[i][j-1]);
			}
		}
	}

	int value = M[len_A][len_B];
	// free memory
	for(i = 0; i<=len_A; i++){
		free(M[i]);
	}
	free(M);
	return value;
}

int main() {
	int t;
	scanf("%d", &t);
	while(t--){
		int len_a, len_b;
		scanf("%d %d", &len_a, &len_b);

		char *a = (char *)calloc(sizeof(char), len_a);
		scanf("%s", a);

		char *b = (char *)calloc(sizeof(char), len_b);
		scanf("%s", b);

		printf("%d\n", longest_common_subsequence(a, b, len_a, len_b));

		free(a);
		free(b);
	}
	return 0;
}
